### PR TITLE
Implementation of the missing CanvasRenderingContext2D functions

### PR DESCRIPTION
**Pull request by luckcolors**
So here it is as mentioned from #43 .
If possible i would like help with reviewing and testing all of the changes ( i'm unfortunately not enough skilled with javascript or gopher js to do that).

I also added the other missing functions and reordered them in categories following Mozila documentation that can be seen [here](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D).

Any feedback will also be greatly appreciated. :)

### CODE REVIEW

**Comment by luckcolors** [comment]
This should be complete right now.

File: dom.go [hunk 1]
```diff
@@ -1946,120 +1992,196 @@ func (e *HTMLCanvasElement) GetContext(param string) *js.Object {
...
+func (ctx *CanvasRenderingContext2D) QuadraticCurveTo(cpx, cpy, x, y float64) {
 	ctx.Call("quadraticCurveTo", cpx, cpy, x, y)
 }
 
-func (ctx *CanvasRenderingContext2D) BezierCurveTo(cp1x, cp1y, cp2x, cp2y, x, y int) {
-	ctx.Call("bezierCurveTo", cp1x, cp1y, cp2x, cp2y, x, y)
-}
-
-func (ctx *CanvasRenderingContext2D) Arc(x, y, r int, sAngle, eAngle float64, counterclockwise bool) {
+func (ctx *CanvasRenderingContext2D) Arc(x, y, r, sAngle, eAngle float64, counterclockwise bool) {
```
Comment (shurcooL) [inline dom.go]: Have you verified that `x`, `y`, and `r` all can be floating point values, making the `int` -> `float64` breaking API change neccessary?

File: dom.go [hunk 2]
```diff
@@ -1946,120 +1992,196 @@ func (e *HTMLCanvasElement) GetContext(param string) *js.Object {
...
 }
 
-// Transformations
+func (ctx *CanvasRenderingContext2D) Rect(x, y, width, height float64) {
+	ctx.Call("rect", x, y, width, height)
+}
 
-func (ctx *CanvasRenderingContext2D) Scale(scaleWidth, scaleHeight int) {
-	ctx.Call("scale", scaleWidth, scaleHeight)
+//Drawing paths
```
Comment (shurcooL) [inline dom.go]: Add a space after `//` here for consistency.

```
// Drawing paths
```

File: dom.go [hunk 3]
```diff
@@ -1946,120 +1992,196 @@ func (e *HTMLCanvasElement) GetContext(param string) *js.Object {
...
-	ctx.Call("clearRect", x, y, width, height)
+// Line styles
+
+func (ctx *CanvasRenderingContext2D) GetLineDash() interface{} {
+	return ctx.Call("getLineDash")
 }
 
-// Paths
+//
+func (ctx *CanvasRenderingContext2D) SetLineDash(arr interface{}) {
```
Comment (shurcooL) [inline dom.go]: What is `arr`? This is pretty unclear.

I see at https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash they're calling the parameter `segments`.

I know this library doesn't set a good precedent for documenting exported identifiers, but it's never too late to start, and this could be a good case where including documentation would be very helpful. How about:

```Go
// SetLineDash sets the line dash pattern used when stroking lines.
// segments is an array of numbers which specify distances to alternately
// draw a line and a gap (in coordinate space units). If the number of
// elements in the array is odd, the elements of the array get copied
// and concatenated. For example, [5, 15, 25] will become [5, 15, 25, 5, 15, 25].
// If the array is empty, the line dash list is cleared and line strokes return to being solid.
func (ctx *CanvasRenderingContext2D) SetLineDash(segments []int) {
}
```

It also looks like we can use `[]int` instead of `interface{}` as the type. Do you agree, or am I missing anything?

File: dom.go (hunk 2 above)
Comment (luckcolors) [inline dom.go]: Yeah most likely just overlooked that one.

File: dom.go (hunk 3 above)
Comment (luckcolors) [inline dom.go]: I put an interface there because i wasn't sure if i could pass arrays into the call function so seeing how gopherjs would convert interface{} to Array wich is what the function wanted i wrote that.
In any case that looks good just be sure to change the type to []float64 this function can work with floats.

File: dom.go (hunk 2 above)
Comment (luckcolors) [inline dom.go]: Yeah missed that one. :D

File: dom.go (hunk 3 above)
Comment (shurcooL) [inline dom.go]: According to https://godoc.org/github.com/gopherjs/gopherjs/js:

```
| Go type               | JavaScript type       | Conversions back to interface{} |
| integers and floats   | Number                | float64
| --------------------- | --------------------- | ------------------------------- |
| []float64             | Float64Array          | []float64                       |
| all other slices      | Array                 | []interface{}                   |
```

So `[]int` would get encoded as `Array` of `Integer`s.

Do we need to allow both `[]int` and `[]float64` as inputs? Then no choice but to use `interface{}`, but we should document the expected types.

Are you sure `float64` inputs are valid? The examples only use integer values. Is it possible to have a segment with a floating point value? Just want to confirm, so we can create the best API possible here.

File: dom.go [hunk 4]
```diff
@@ -2072,6 +2194,11 @@ func (ctx *CanvasRenderingContext2D) Restore() {
 	ctx.Call("restore")
 }
 
+//Hit regions
+//addHitRegion
+//removeHitRegion
+//clearHitRegions
```
Comment (luckcolors) [inline dom.go]: Yeah i meant to put a TODO list there, i was using comments like this to note down wich functions i had to implement and where they had to go.

The only reason i didn't implement these is that those require the https://developer.mozilla.org/en-US/docs/Web/API/Path2D

If you think we need those i could implement those for another pull request.

Comment (shurcooL) [inline dom.go]: It's completely okay to leave them as a TODO for the future.

You should only implement them if you want to or need them yourself. I don't need them right now.

Either way, it's better to leave out for another PR, to avoid making this one larger.

File: dom.go (hunk 1 above)
Comment (luckcolors) [inline dom.go]: Coordinates can be float64 the mantissa part does actually influence when using drawinf functions.

The only case where i could see the float64 being a problem is in functions that work with array indexes.

That last statement might also be actually false i just tested with DrawImage PutImageData, and GetImageData and it works without issue with floats.

Comment (luckcolors) [inline dom.go]: CreateImageData is probably the only function on wich it would make sense to have ints, then again if we have all of the canvas functions work with float64 then we could just leave them as it is.

About this topic https://stackoverflow.com/questions/6649283/how-are-floating-point-array-indices-interpreted

File: dom.go [hunk 5]
```diff
@@ -1946,120 +1992,196 @@ func (e *HTMLCanvasElement) GetContext(param string) *js.Object {
...
+}
 
-	ctx.Call("strokeText", text, x, y, maxWidth)
+func (ctx *CanvasRenderingContext2D) DrawImageWithSrcAndDst(image *Element, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight float64) {
+	ctx.Call("drawImage", image, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight)
+}
+
+// Pixel manipulation
+
+func (ctx *CanvasRenderingContext2D) CreateImageData(width, height float64) *ImageData {
```
Comment (shurcooL) [inline dom.go]: I think we should use `int`s for indexes and dimensions. ImageData represents pixel data, and pixels ("picture elements") are distinct elements. As far as I can tell, there can't be valid floating point values for width and height, etc.

image

File: dom.go (hunk 3 above)
Comment (luckcolors) [inline dom.go]: Yeah they are. I mean if you use the sanbox on the page and put say 50.5 you can actually see the image changing.

File: dom.go (hunk 5 above)
Comment (luckcolors) [inline dom.go]: I agree.

File: dom.go (hunk 4 above)
Comment (luckcolors) [inline dom.go]: Sure thing.

**Comment by luckcolors** [comment]
@dominikh Ping :)

**Comment by dominikh** [comment]
@shurcooL ping.

**Comment by shurcooL** [comment]
@luckcolors You haven't applied the changes from my last review on August 22 yet. You said in comments you agreed, but there haven't been any new commits since.

File: dom.go (hunk 3 above)
Comment (shurcooL) [reply dom.go]: If we're sticking with `interface{}` argument here, we definitely need documentation to explain how to use it. See first comment for a starting point.

File: dom.go (hunk 2 above)
Comment (shurcooL) [reply dom.go]: This isn't done yet.

File: dom.go (hunk 5 above)
Comment (shurcooL) [reply dom.go]: This isn't done yet.

**Comment by luckcolors** [comment]
AH, sorry this is my first time doing a code review, i thought the "reviews" where atuomatically applied to the code.
I will push the changes as soon as possible.
:)

**Comment by luckcolors** [comment]
While i'm at this again if you want i can figure out the functions i left in the todo.

File: dom.go (hunk 1 above)
Comment (luckcolors) [reply dom.go]: I changed in CreateImageData the type from float64 to int.

File: dom.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (luckcolors) [reply dom.go]: Fixed the TODO there.

File: dom.go (hunk 3 above)
Comment (luckcolors) [reply dom.go]: Ok so i went with []float64, it seems just easier to use anyway, as the function does work with both value types, interface{} seems an exageration as we can easily fit int values in floats.

File: dom.go (hunk 5 above)
Comment (luckcolors) [reply dom.go]: Changed to int.

File: dom.go (hunk 2 above)
Comment (luckcolors) [reply dom.go]: Fixed the typo.

**Comment by shurcooL** [comment]
> While i'm at this again if you want i can figure out the functions i left in the todo.

Feel free to decide for yourself how much work you're willing to take on. I would suggest leaving those out for a separate PR, if you want.

File: dom.go (hunk 1 above)
Comment (shurcooL) [reply dom.go]: What about `GetImageData`, `PutImageData` and `PutImageDataDirty`, some of those also involve widths/heights. Should they be changed to `int`? If not, what's the reason to keep them `float64`?

Comment (luckcolors) [reply dom.go]: Ok i've checked all three, ```GetImageData``` doesn't suffer any changes and works with floating values,
```PutImageData``` does work with floats and it has a graphical difference (same goes with ```PutImageDataDirty``` for all of the parameters).


Also to put this question to an end: https://stackoverflow.com/questions/29371695/does-canvas-support-floating-point-numbers-when-drawing-line-or-arc

I also noticed i left ```LineTo``` ```MoveTo``` ```BezierCurveTo``` with int, i changed them to float64 as those work aswell with floating values. (I verified that)

Comment (luckcolors) [reply dom.go]: Another reason would be ease of use, i agree with CreateImageData that it's a good idea to use int,
(like ease of use of len as an input).
All the others do work with float64 and also it makes more sense to keep them with only one datatype.

So i think this is now done, unless you have some other doubts or issues you would like to address.

File: dom.go [hunk 6]
```diff
@@ -1946,120 +1992,196 @@ func (e *HTMLCanvasElement) GetContext(param string) *js.Object {
...
+	textMetrics := ctx.Call("measureText", text)
+	return &TextMetrics{Object: textMetrics}
 }
 
-func (ctx *CanvasRenderingContext2D) ClearRect(x, y, width, height int) {
-	ctx.Call("clearRect", x, y, width, height)
+// Line styles
+
+func (ctx *CanvasRenderingContext2D) GetLineDash() []float64 {
+	return ctx.Call("getLineDash")
```
Comment (shurcooL) [inline dom.go]: Did you test these changes? It doesn't build:

```
$ gopherjs build
dom.go:2036:9: cannot use ctx.Call("getLineDash") (value of type *github.com/gopherjs/gopherjs/js.Object) as []float64 value in return statement
```

**Comment by shurcooL** [comment]
> So i think this is now done, unless you have some other doubts or issues you would like to address.

I think this is almost ready to be merged. There just 2 known issues we need to resolve:

1. `GetLineDash()` doesn't build. See inline comment above. We need to fix that.
2. I'm still thinking whether it's better to use ints or floats for `GetImageData`... If I don't have a conclusive answer on this, we can go with what you've chosen as is, and if there are good reasons, we can still change this later. People will be okay with breaking API changes here since it's a new API, and we can't always know the best long term solution before trying it.

Aside from that, I don't see any other issues. Thanks a lot for working on this!

**Comment by shurcooL** [comment]
I've pushed a fix for `GetLineDash` build error in df6ab95b9c4396bfdd1c300bca4992016d1ce894, please review @luckcolors.

**Comment by shurcooL** [comment]
I've done some reading of https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData and testing `GetImageData`:

```Go
println(ctx.GetImageData(50, 50, 1, 1).Object)
// Output: ImageData {data: Uint8ClampedArray(4), width: 1, height: 1}

println(ctx.GetImageData(50, 50, 1.999, 1).Object)
// Output: ImageData {data: Uint8ClampedArray(4), width: 1, height: 1}

println(ctx.GetImageData(50, 50, 0, 1).Object)
// Output: Failed to execute 'getImageData' on 'CanvasRenderingContext2D': The source width is 0.

println(ctx.GetImageData(50, 50, 0.999, 1).Object)
// Output: Failed to execute 'getImageData' on 'CanvasRenderingContext2D': The source width is 0.
```

It becomes quite clear that the float parameters are truncated to integers.

These are pixel-manipulation methods. They operate on pixels. A single pixel in `ImageData` is a 4 `uint8` values (R, G, B, A components). It doesn't makes sense to have floats when it comes to precise pixel manipulation, so I'm very confident we must use `int` here.

Also, `ImageData` width/height properties are `int`:

```Go
type ImageData struct {
	*js.Object
	Width  int        `js:"width"`
	Height int        `js:"height"`
```

Not `float64`.

However, `drawImage` is different. It doesn't operate on exact pixels, and can accept fractional values. I tested by using a browser zoom of 500%. At that point, each pixel was actually 5 real pixels wide, and fractional values to `drawImage` had a visible effect.

In conclusion, I'll change all pixel manipulation methods to use `int` parameters, since they operate on exact pixel data and floating points don't make sense. But the rest are good with `float64`.

**Comment by shurcooL** [comment]
I testes the X/Y positions, and can confirm they're truncated as well as width/height:

```Go
var canvas = document.GetElementByID("canvas").(*dom.HTMLCanvasElement)
var ctx = canvas.GetContext2d()

ctx.FillStyle = "rgb(12, 34, 45)"
ctx.Rect(0, 0, 100, 100)
ctx.Fill()

println(ctx.GetImageData(50, 99, 1, 1).Data)
println(ctx.GetImageData(50, 99.999, 1, 1).Data)
println(ctx.GetImageData(50, 100, 1, 1).Data)

// Output:
// Uint8ClampedArray(4) [12, 34, 45, 255]
// Uint8ClampedArray(4) [12, 34, 45, 255]
// Uint8ClampedArray(4) [0, 0, 0, 0]
```

**Comment by shurcooL** [comment]
`putImageData` is similar to `drawImage`, in that it accepts fractional values for the position X/Y of where in the canvas to render image data. However, the dirty parameters can only be integers, since they refer to the pixel-based nature of the `ImageData`.

**Comment by luckcolors** [comment]
I did some more testing with https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData and i checked, the dirtyX and dirtyY do change based on the floating part,
dirtyWidth and dirtHeight get truncated if a floating part exist (as you also verified with getImageData).

**Comment by luckcolors** [comment]
> I've reviewed the rest of this PR and addressed minor issues I found. This has my LGTM now.

Soo no problem! Thanks to you for your time and patience.
This has also my LGTM now, (thanks for spotting those tiny errors).
👍

**Comment by shurcooL** [comment]
Sure thing, happy to collaborate! Merged. Thank you again @luckcolors!
